### PR TITLE
Tic tac toe - React App

### PR DESCRIPTION
**Pull request by fededirocco**
## Summary
* Implemented tic-tac-toe game with turns and past move list.
* Added Board, Game and Square components. 
* Modified App component calling Game component.
* Used CSS Modules for SCSS.

## Trello Card
https://trello.com/c/G8OE46yu/43-react-app-tic-tac-toe

## Screenshots
tictactoe

### CODE REVIEW

**Comment by carlosbollero** [comment]
When the app gains functionality it would be nice to, instead of screenshots, have gifs showing the app live.
On Ubuntu, [Peek](https://github.com/phw/peek#ubuntu) is a fine alternative.

File: training-react/src/app/components/Game/index.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    return null;
+  };
+
+  render() {
+    const history = this.state.history;
+    const current = history[this.state.stepNumber];
+    const winner = this.calculateWinner(current.squares);
+
+    const moves = history.map((step, move) => {
```
[the remark this answers is not in the document]
Comment (fededirocco) [reply training-react/src/app/components/Game/index.js]: I tried to move the desc constant to a function and then call it inside render so for every desc i can render his button. But I couldn't, any help?

[the remark this answers is not in the document]
Comment (fededirocco) [reply training-react/src/app/components/Game/index.js]: I moved the implementation to get status outside render.

File: training-react/src/app/components/Game/index.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    } else {
+      status = 'Next player: ' + (this.state.xIsNext ? 'X' : 'O');
+    }
+
+    return (
+      <div className={styles.game}>
+        <div className={styles.board}>
+        <Board
+          squares={current.squares}
+          onClick={(i) => this.handleClick(i)}
```
[the remark this answers is not in the document]
Comment (fededirocco) [reply training-react/src/app/components/Game/index.js]: I think we need to know the {i} for every square. Did you see a solution for this?

File: training-react/src/app/components/Game/index.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        'Go to move #' + move :
+        'Go to game start';
+      return (
+        <li key={step}>
+          <button onClick={() => this.jumpTo(move)}>{desc}</button>
+        </li>
+      );
+    });
+
+    let status;
```
[the remark this answers is not in the document]
Comment (fededirocco) [reply training-react/src/app/components/Game/index.js]: If I put a const, I have an error

File: training-react/src/app/components/Game/index.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      xIsNext: true,
+      stepNumber: 0
+    };
+  }
+
+  handleClick = i => {
+    const history = this.state.history.slice(0, this.state.stepNumber + 1);
+    const current = history[history.length - 1];
+    const squares = current.squares.slice();
+    if (this.calculateWinner(squares) || squares[i]) {
```
[the remark this answers is not in the document]
Comment (fededirocco) [reply training-react/src/app/components/Game/index.js]: I implemented a solution (winner as a state property) but its has a problem when it has to set the state when is a winner.

File: training-react/src/app/components/Game/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (carlosbollero) [reply training-react/src/app/components/Game/index.js]: What if you do this `onClick={this.handleClick}`?

File: training-react/src/app/components/Game/index.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    return status;
+  };
+
+  handleClick = i => {
+    if (!this.props.isWinner) {
+      const history = this.state.history.slice(0, this.state.stepNumber + 1);
+      const current = history[history.length - 1];
+      const squares = [...current.squares];
+      if (calculateWinner(squares) || squares[i]) {
```
Comment (carlosbollero) [inline training-react/src/app/components/Game/index.js]: As you already know there is no winner, i think it should be enough to check if the square is filled.
What if we just check `if (!squares[i])` and set the winner inside setState?
Let me know

File: training-react/src/app/components/Game/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (fededirocco) [reply training-react/src/app/components/Game/index.js]: Yes, its work! Now i understand because the {i} is in every square

**Pull request merged**